main/activity-log.js

```diff
@@ -0,0 +1,79 @@
+'use strict'
+
+const Store = require('electron-store')
+const configStore = new Store({
+  name: 'activity-log'
+})
+
+class ActivityLog {
+  #entries
+  #lastId
+
+  constructor () {
+    this.#entries = loadStoredEntries()
+    this.#lastId = this.#entries
+      // We are storing ids as strings to allow us to switch to GUIDs in the future if needed
+      // When looking for the max id used, we need to cast strings to numbers.
```

Comment:
also, shouldn't the highest id always belong to the last entry, so no search is necessary?

Reply:
Yes, I am worried about collisions.

I guess you are right that we can simply use id of the last item as the last id 👍🏻 

> do we even needs ids? if the position is enough, we can address by array index

There is a race condition in the UI:
1. You want to subscribe for new activity entries  
2. You also want to fetch the list of existing entries. This is an asynchronous operation.

I think the correct implementation is this:
1. Subscribe to new activity
2. Get the existing activity

Now it's possible that an activity is recorded with such a timing that this entry triggers your event listener (1) and is also returned for your RPC call (2). The id allows the frontend to de-duplicate data coming from these two sources.

Reply:
See [8a759de](https://github.com/filecoin-project/filecoin-station/pull/110/commits/8a759de1fdff4643442bf8e258ba49d0914d2789)

Reply:
Makes sense!

> Yes, I am worried about collisions.

https://www.npmjs.com/package/monotonic-timestamp comes to mind for this, but I'm not sure if that actually improves our situation, as it tries to solve what we currently do with `.id` and `.timestamp` in one value.

> There is a race condition in the UI:

We might be able to fix this on the backend. Say the client only subscribes, never manually fetches the historical records. When the backend receives a subscription, it also pushes historical data over it. As long as the backend can keep items in sync, we wouldn't need IDs this way and could simply use array position.

Reply:
> Say the client only subscribes, never manually fetches the historical records. When the backend receives a subscription, it also pushes historical data over it.

That's a cool design. 💯👍🏻 It will simplify frontend code a lot.

Do you know how to implement the part about detecting a subscription? We are using Electron IPC and Node.js EventEmitter APIs, I don't remember seeing a straightforward way for detecting new event subscriptions.

A solution that comes to my mind:
- Keep the current design where the frontend registers an event listener first and then sends an RPC call to the backend.
- By default, backend does not emit any events - it's in _paused_ state.
- Rework the RPC call from "get entries" to ask the backend to start emitting the activity entries instead. (Resume the stream.)

What do you think, @juliangruber? Do you have any other ideas on how to implement this part?

Reply:
> do we even needs ids? if the position is enough, we can address by array index

When the Activity Log reaches the max size, we remove the oldest entry from the array. This changes the index of all entries from `X` to `X-1`. Each new event gets the same index `99`.

Reply:
Reworked activity streaming from `main` to `renderer` based on my description above - see 7e2dcad05

Reply:
> What do you think, @juliangruber? Do you have any other ideas on how to implement this part?

I thought of it the same way, I think this will work well!

> When the Activity Log reaches the max size, we remove the oldest entry from the array. This changes the index of all entries from X to X-1. Each new event gets the same index 99.

With this new design, what do we even need IDs for? If we need them for the react `key={id}` prop, then we could assign random IDs. I don't see the need for monotonically increasing IDs with this design 🤔

Reply:
> With this new design, what do we even need IDs for? If we need them for the react `key={id}` prop, then we could assign random IDs. I don't see the need for monotonically increasing IDs with this design 🤔

Yes, I was thinking about `key={id}` prop.

Does it actually matter how we are generating the id? What are your concerns about storing the `id` at backend?

Reply:
I'm absolutely for storing the id at the backend, but I think they can be random values and don't need to be monotonically increasing, since we don't need to use them for sorting any more, right? We can sort by timestamp, and iff two values have the same timestamp we could fall back to sorting these by their id. I think the system will be easier to reason about if ids are only used for identifying, and timestamps are used for sorting. Unless I miss something, in which case we could also have a quick call about this?

Reply:
is uuids an option for random ids? I think it's useful to have unique id for activity log entries.
for the UI, unique ids simplifies things (eg `key={id}`) 👍 
timestamps can help with ordering (although we can just read whatever order comes from backend).

Reply:
To help resolve this PR, I've made a PR demonstrating how I think this could work in a simpler way: https://github.com/filecoin-project/filecoin-station/pull/127